### PR TITLE
os/fs/driver/mtd: Fix for CRC check during smartfs-image host build

### PR DESCRIPTION
**Pull request by amogh-samsung**
-Separated CRC check modifications for host build
 from on-device functioning usinf NXFUSE_HOST_BUILD flag.
-Avoids unneccessary erase of eraseblocks due to CRC
 invalidation of completely erased logical sectors during
 host build.

Signed-off-by: Amogh Hassija <a.hassija@samsung.com>

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [8ec447b1b5fe8ea80470610e5559e4bcf566f1a0] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [8ec447b1b5fe8ea80470610e5559e4bcf566f1a0] - Code Rule Check (C++) OK.

File: os/fs/driver/mtd/smart.c [hunk 1]
```diff
@@ -4049,6 +4051,7 @@ static int smart_write_alloc_sector(FAR struct smart_struct_s *dev, uint16_t log
  *
  ****************************************************************************/
 
+#ifdef NXFUSE_HOST_BUILD
```
Comment (Taejun-Kwon) [inline os/fs/driver/mtd/smart.c]: you don't have to make 2 functions, I think.
You'd better modify smart_scan itself to apply extra logic for sector '0'

File: os/fs/driver/mtd/smart.c [hunk 2]
```diff
@@ -4059,30 +4062,67 @@ static int smart_validate_crc(FAR struct smart_struct_s *dev)
...
-#else
-	/* Test 8-bit CRC for CRC8 & basic case for smart_scan. */
+#elif defined(CONFIG_SMART_CRC_8)
+	/* Test 8-bit CRC for CRC8. */
+
 	if (crc != header->crc8 && crc != SMART_ERASEDSTATE_CRC) {
 		return -EIO;
 	}
+#else
+	/*CONFIG_MTD_SMART_ENABLE_CRC not set, perform minimal CRC check
```
Comment (Taejun-Kwon) [inline os/fs/driver/mtd/smart.c]: As I told you before,  below can't be acceptable
```
crc != SMART_ERASEDSTATE_CRC 
```
Bit flip can be occured during the power down case.

File: os/fs/driver/mtd/smart.c (hunk 1 above)
Comment (amogh-samsung) [reply os/fs/driver/mtd/smart.c]: HI Mr. Kwon,
The repeated definition is for the host-build case (NXFUSE_HOST_BUILD) where smartfs.bin image is being pre-built on the Linux(host) machine for flashing to the device. 
It separates the logic from what would be involved during on device smartfs operations.

File: os/fs/driver/mtd/smart.c (hunk 2 above)
Comment (amogh-samsung) [reply os/fs/driver/mtd/smart.c]: - Yes sir, agreed, with reference to your inputs on the previous PR, the functioning of the CRC check (smart_validate_crc function) and the **hard coding for the pre-build of smartfs.bin has been separated from on-device functioning** by use of NXFUSE_HOST_BUILD flag.
- The hard coded values are **used only when smartfs.bin image is being created on the host(Linux) machine** where abrupt power outages may not pose a problem.
- SMART_ERASEDSTATE_CRC **does not come into play at all** when smartfs operations are being carried out **on the device.**

**Comment by seinfra** [comment]
Target : [e01dcbf9ab322c055221a87bf7b38b9a9a9925c2] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e01dcbf9ab322c055221a87bf7b38b9a9a9925c2] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [8ec447b1b5fe8ea80470610e5559e4bcf566f1a0] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [8ec447b1b5fe8ea80470610e5559e4bcf566f1a0] - Code Rule Check OK.

**Pull request merged**